subsys/power/power.c

```diff
@@ -121,11 +122,12 @@ enum power_states sys_suspend(s32_t ticks)
 	sys_set_power_state(pm_state);
 	sys_pm_debug_stop_timer();
 
+#if CONFIG_DEVICE_POWER_MANAGEMENT
 	if (deep_sleep) {
```

Comment:
There was (and still is) a link between device and system power management. When you are entering DEEP_STATE mode, the system power management shutdown the devices.

Please note, that this part is far (really far) from perfect. We will be able to make it better when real device power management will be available.

Reply:
A tighter integration between device and system power management definitely makes a lot of sense. Especially when looking into the future. The issue here is a different one.

When we call `sys_set_power_state()` to enter deep sleep mode we loose execution context. I.e. to wake up we will need to do system reset. The next line of code executed will be reset handler and not the `if (deep_sleep) {` statement.

Reply:
The sys_resume() should handle that. However I am not sure if other things are ready for such scenario. When device PM will be there I will try to use RAM retention and have deep sleep with context saving working.

Reply:
> The sys_resume() should handle that.

`sys_resume()` will not be executed after waking up from the deep sleep. The next line of code executed will be in the reset handler.

> When device PM will be there I will try to use RAM retention and have deep sleep with context saving working.

Currently the description of Deep Sleep State mentions among its characteristics: "RAM is selectively retained". That's however more of a special case. The description should be updated. While there are probably SoCs that offer partial RAM retention in Deep Sleep mode that's definitely not a standard feature.

Reply:
The lack of proper description of power modes is a real problem.
According to code, you have 3 ways to exit deep sleep:

1) Continue execution. RAM retention is assumed.
2) Through sys_resume(): Low-level code will call this function from ISR, and then normal execution will continue. RAM retention assumed in this case.
3) Through reset. No RAM retention is assumed.

Please note, that so far I not touched low-level part of the power management. All the changes are focusing on policies and power state naming. To get into low-level stuff we need proper device power management, which is under development. When it will be done, I will get back to system-level management and I will try to make it really working.

Up to the time feel free to propose a good abstraction model of system-level power management, better description of power modes etc. There is still a lot of not-so-defined things in the area.

Reply:
> According to code, you have 3 ways to exit deep sleep:
> 1. Continue execution. RAM retention is assumed.
> 2. Through sys_resume(): Low-level code will call this function from ISR, and then normal execution will continue. RAM retention assumed in this case.
> 3. Through reset. No RAM retention is assumed.

The first two options would only be required if there exists a SoC where in the Deep Sleep mode the CPU has full RAM retention but all the peripherals are powered off. I am not aware of any. But maybe I was to quick to make a judgment and we should cover also such scenario. In a typical case Deep Sleep mode means no (possibly partial) RAM retention and exit via a system reset.

There may be some confusion around vendors like Intel. On Intel Quark microcontroller there are power states called deep sleep where the state of the CPU core as well as other SoC components are preserved and exit takes a few us. That's what an ARM vendor calls a low power mode.